Feature: Allow variants to propagate if not available in source pkg

RikkiButler20 opened the pull request:
Fixes #34302.

Variants can now be propagated from a dependent package to (transitive) dependencies, 
even if the source or transitive dependencies have the propagated variants.

For example, here `zlib` doesn't have a `guile` variant, but `gmake` does:
```
$ spack spec zlib++guile
 -   zlib@1.3%gcc@12.2.0+optimize+pic+shared build_system=makefile arch=linux-rhel8-broadwell
 -       ^gcc-runtime@12.2.0%gcc@12.2.0 build_system=generic arch=linux-rhel8-broadwell
 -       ^gmake@4.4.1%gcc@12.2.0+guile build_system=generic arch=linux-rhel8-broadwell
```

Adding this property has some strange ramifications for `satisfies()`. In particular:
* The abstract specs `pkg++variant` and `pkg+variant`  do not intersect, because `+variant`
  implies that `pkg` *has* the variant, but `++variant` does not.
* This means that `spec.satisfies("++foo")` is `True` if:
    * for concrete specs: `spec` and its dependencies all have `foo` set if it exists
    * for abstract specs: no dependency of `spec`  has `~foo` (i.e. no dependency contradicts `++foo`).
* This also means that `Spec("++foo").satisfies("+foo")` is `False` -- we only know after concretization.

The `satisfies()` semantics may be surprising, but this is the cost of introducing non-subset
semantics (which are more useful than proper subsets here).

- [x] Change checks for variants
- [x] Resolve conflicts
- [x] Add tests
- [x] Add documentation

RikkiButler20 committed:
Add test

RikkiButler20 committed:
Remove conflict & Skip variant check
Don't check in _spec_clause if variant is found in pkg

RikkiButler20 committed:
Add test for propagating one variant and not another
Fix style

RikkiButler20 commented:
@spackbot Fix style

RikkiButler20 committed:
[@spackbot] updating style on behalf of RikkiButler20

RikkiButler20 committed:
Update propgation call

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 committed:
Merge remote-tracking branch 'upstream/develop' into feature/propagate-variant-not-source

RikkiButler20 committed:
Update test for variat propagation
Add comments with notes for feature

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 commented:
@spackbot Fix style

RikkiButler20 committed:
[@spackbot] updating style on behalf of RikkiButler20

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 committed:
Add more propagation test

RikkiButler20 commented:
@spackbot Fix style

RikkiButler20 committed:
[@spackbot] updating style on behalf of RikkiButler20

RikkiButler20 committed:
Fix style & add comments to tests

RikkiButler20 committed:
Add test & change mock pkgs
Test for propagating variant that's not in source to multiple deps

RikkiButler20 committed:
If variant is propagatin it satisfies (constraints are too loose)

RikkiButler20 committed:
Satisfies returns the variants that are propagating

RikkiButler20 commented:
@spackbot Fix style

RikkiButler20 committed:
[@spackbot] updating style on behalf of RikkiButler20

RikkiButler20 committed:
Update test
No more exclusions allowed for variant propagation

RikkiButler20 committed:
Stop allowing exclusions when propagating variants
It looks terrible

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 committed:
Add test for propagating variants past 1st level deps

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 committed:
Check for propagation in variant.py

tgamblin commented:
@spackbot Fix style

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 commented:
@spackbot Fix style

RikkiButler20 committed:
Add new integrity constraints

becker33 committed:
add more internal_error facts to solver

becker33 committed:
TODO note

RikkiButler20 committed:
Only set variant if pkg has variant

RikkiButler20 committed:
Add test for propagating multiple variants

RikkiButler20 committed:
Add test for single valued variants

RikkiButler20 committed:
Fix spelling

RikkiButler20 committed:
Add test for propagating variants not in source pkg

RikkiButler20 committed:
Add error message for excluding a propagating variant
Does not look good. Fix message later

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 commented:
@spackbot Fix style

RikkiButler20 committed:
[@spackbot] updating style on behalf of RikkiButler20

RikkiButler20 commented:
@spackbot Fix style

RikkiButler20 committed:
Fix style

RikkiButler20 committed:
Fix satisfies constraints
Check that variant is in pkg's dependencies in a separate method

RikkiButler20 committed:
Fix style

RikkiButler20 committed:
Fix style & add comments

RikkiButler20 committed:
Add back missing test

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 committed:
Allow multivalued variants to include a second value when propagatin

RikkiButler20 commented:
@spackbot Fix style

RikkiButler20 committed:
[@spackbot] updating style on behalf of RikkiButler20

RikkiButler20 committed:
Fix style
spack bot keeps changing this

RikkiButler20 committed:
Clean up code
Re-add test again

RikkiButler20 committed:
Style

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 committed:
Add some info about propagation to docs

RikkiButler20 committed:
Fix variant_value check

RikkiButler20 committed:
Make code more readable

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 committed:
Make error message more readable

RikkiButler20 committed:
Find source package for error message

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1077,6 +1088,30 @@ variant_is_propagated(PackageNode, Variant) :-
+  not variant_type(VariantID, "multi"),
+
+  % Find the source based on the fact that nothing is propagating to it
+  propagate(node(Y, Source), variant_value(Variant, Value2)),
+  not 1 {propagate(SourceParent, variant_value(Variant, Value2)): depends_on(SourceParent, node(Y, Source))}.
```
Curious what @alalazo thinks about this. I wonder if it would be too expensive to pull this out and use it elsewhere, e.g.:

```suggestion
  propagation_source(node(Y, Source), variant_value(Variant, Value2)).
```

and, elsewhere:

```prolog
% a node is the source of propagation if nothing propagated to it
propagation_source(Source, PropagatedValue) :-
  propagate(Source, PropagatedValue),
  not 1 { propagate(Parent, PropagatedValue): depends_on(Parent, Source) }.
```

This is a little clumsy because the `propagate` rules don't preserve any information on what the source was. So we're deriving something we already knew (but forgot)  We could tweak the encoding in `asp.py` so that the `asp.py` just sets `propagate_source` and the solver derives `propagate` -- them we wouldn't have to be so fancy.

Thoughts @alalazo @becker33?

alalazo replied:
I think my answer would basically be https://github.com/spack/spack/pull/42931#discussion_r1823254801 We should try to get the second argument in `propagate/2` contain info on the source node.

alalazo replied:
So, it seems this might be causing a performance issue after all, see https://github.com/spack/spack/pull/42931#issuecomment-2449224921

alalazo replied:
I'm going to push the diff below, since it passes tests locally, and it performs better when grounding and solving. The gap between the two widens the more variants you propagate. As an example, with the patch:
```console
$ spack solve --timers trilinos++aggressive_vectorization++shared++debug++mpi
    setup         14.512s
    load           2.219s
    ground         7.778s
    solve          6.861s
    total         32.089s

[ ... ]
```
without the patch:
```console
$ spack solve --timers trilinos++aggressive_vectorization++shared++debug++mpi
    setup         14.511s
    load           2.247s
    ground         9.031s
    solve          7.600s
    total         34.283s

 [ ... ]
```

alalazo replied:
Fixed in https://github.com/spack/spack/pull/42931/commits/61b3ba3ceefbd914caf76af6d484816f313dcd0f

RikkiButler20 committed:
Fix error message for propagating variant twice
It shows the error twice

RikkiButler20 committed:
Can propagate variant from diff source pkg
If the variant is not singlevalued

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 committed:
Don't allow propagating variants from multiple source pkgs
if one of the pkgs is a direct dependency of another

RikkiButler20 committed:
Error propagating variant in diamond dependency

RikkiButler20 committed:
Create propagation_source rule

RikkiButler20 committed:
Merge branch 'develop' into feature/propagate-variant-not-source

RikkiButler20 committed:
Add test & errors for propagating variants with diamond dep

RikkiButler20 committed:
Merge branch 'feature/propagate-variant-not-source' of github.com:RikkiButler20/spack into feature/propagate-variant-not-source

RikkiButler20 committed:
Check all dependencies when checking for error when propagating

RikkiButler20 committed:
Stop error from showing up twice

RikkiButler20 committed:
Make test's assertions more explicit

RikkiButler20 committed:
Fix style & comments

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2028,9 +2028,12 @@ def _spec_clauses(
-                clauses.append(f.variant_value(spec.name, vname, value))
                 if variant.propagate:
                     clauses.append(f.propagate(spec.name, fn.variant_value(vname, value)))
+                    if vname in self.pkg_class(spec.name).variants:
```
This is incorrect since #44425 got merged:
```suggestion
                    if self.pkg_class(spec.name).has_variant(vname):
```
Unfortunately fixing this seems to make `test_concretize_propagate_multiple_multivalue_variant` FAIL.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1077,6 +1088,54 @@ variant_is_propagated(PackageNode, Variant) :-
    not propagate(PackageNode, variant_value(Variant, Value)).
 
+% a node is the source of propagation if nothing propagated to it
+propagation_source(Source, Variant, Value) :-
+  propagate(Source, variant_value(Variant, Value)),
+  not 1 { propagate(Parent, variant_value(Variant, Value)): depends_on(Parent, Source) }.
```
(Not necessarily for this PR) I think this can be improved, in the sense that we know from user input which are the sources of propagation, and we don't need to have cardinality conditions taking into account all possible deps to recompute that here.

alalazo replied:
https://github.com/spack/spack/pull/42931/commits/61b3ba3ceefbd914caf76af6d484816f313dcd0f

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1077,6 +1088,54 @@ variant_is_propagated(PackageNode, Variant) :-
+error(100, "{0} and {1} cannot both propagate variant '{2}' to the shared dependency: {3}",
+  Package1, Package2, Variant, Dependency) :-
+  % Package1 and Package2 are source packages set to propagate Variant
+  propagation_source(node(X, Package1), Variant, Value1),
+  propagation_source(node(Y, Package2), Variant, Value2),
+  % Package1 and Package2 and their values are different
+  Package1 > Package2,
+  Value1 != Value2,
+  % The variant is a singlevalued variant
+  variant_single_value(node(X, Package1), Variant),
+  % Dependency is trying to propagate Variant with different values and is not the source package
+  propagate(node(Z, Dependency), variant_value(Variant, Value1)),
+  propagate(node(Z, Dependency), variant_value(Variant, Value2)),
```
Another possible improvement here would be to put the source node within the propagated attribute:
```
propagate(node(Z, Dependency), variant_value(Variant, Value1, SourceNode)),
```
That way we don't have to add `propagate_source/3` to the encoding.

alalazo replied:
This would be something similar to:
```prolog
non_default_propagation(variant_value(Name, Value)) :- attr("propagate", RootNode, variant_value(Name, Value)).

% Propagation roots have a corresponding attr("propagate", ...)
propagate(RootNode, PropagatedAttribute) :- attr("propagate", RootNode, PropagatedAttribute), not non_default_propagation(PropagatedAttribute).
propagate(RootNode, PropagatedAttribute, EdgeTypes) :- attr("propagate", RootNode, PropagatedAttribute, EdgeTypes).

% Special case variants, to inject the source node in the propagated attribute
propagate(RootNode, variant_value(Name, Value, RootNode)) :- attr("propagate", RootNode, variant_value(Name, Value)).

```

alalazo replied:
@tgamblin @RikkiButler20 This one seems to work, modulo the failing test above that keeps failing, and removes `propagation_source/3`:
```diff
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index a2298c9f0a..fc1499341f 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -990,7 +990,7 @@ variant_not_default(node(ID, Package), Variant, Value)
     % variants set explicitly on the CLI don't count as non-default
     not attr("variant_set", node(ID, Package), Variant, Value),
     % variant values forced by propagation don't count as non-default
-    not propagate(node(ID, Package), variant_value(Variant, Value)),
+    not propagate(node(ID, Package), variant_value(Variant, Value, _)),
     % variants set on externals that we could use don't count as non-default
     % this makes spack prefer to use an external over rebuilding with the
     % default configuration
@@ -1002,7 +1002,7 @@ variant_default_not_used(node(ID, Package), Variant, Value)
   :- variant_default_value(node(ID, Package), Variant, Value),
      node_has_variant(node(ID, Package), Variant, _),
      not attr("variant_value", node(ID, Package), Variant, Value),
-     not propagate(node(ID, Package), variant_value(Variant, _)),
+     not propagate(node(ID, Package), variant_value(Variant, _, _)),
      attr("node", node(ID, Package)).
 
 % The variant is set in an external spec
@@ -1047,10 +1047,14 @@ variant_single_value(PackageNode, Variant)
 % Propagation semantics
 %-----------------------------------------------------------------------------
 
+non_default_propagation(variant_value(Name, Value)) :- attr("propagate", RootNode, variant_value(Name, Value)).
+
 % Propagation roots have a corresponding attr("propagate", ...)
-propagate(RootNode, PropagatedAttribute) :- attr("propagate", RootNode, PropagatedAttribute).
+propagate(RootNode, PropagatedAttribute) :- attr("propagate", RootNode, PropagatedAttribute), not non_default_propagation(PropagatedAttribute).
 propagate(RootNode, PropagatedAttribute, EdgeTypes) :- attr("propagate", RootNode, PropagatedAttribute, EdgeTypes).
 
+% Special case variants, to inject the source node in the propagated attribute
+propagate(RootNode, variant_value(Name, Value, RootNode)) :- attr("propagate", RootNode, variant_value(Name, Value)).
 
 % Propagate an attribute along edges to child nodes
 propagate(ChildNode, PropagatedAttribute) :-
@@ -1072,7 +1076,7 @@ propagate(ChildNode, PropagatedAttribute, edge_types(DepType1, DepType2)) :-
 
 % If a variant is propagated, and can be accepted, set its value
 attr("variant_selected", PackageNode, Variant, Value, VariantType, VariantID) :-
-  propagate(PackageNode, variant_value(Variant, Value)),
+  propagate(PackageNode, variant_value(Variant, Value, _)),
   node_has_variant(PackageNode, Variant, VariantID),
   variant_type(VariantID, VariantType),
   variant_possible_value(PackageNode, Variant, Value),
@@ -1081,60 +1085,45 @@ attr("variant_selected", PackageNode, Variant, Value, VariantType, VariantID) :-
 % If a variant is propagated, we cannot have extraneous values
 variant_is_propagated(PackageNode, Variant) :-
   attr("variant_value", PackageNode, Variant, Value),
-  propagate(PackageNode, variant_value(Variant, Value)),
+  propagate(PackageNode, variant_value(Variant, Value, _)),
   not attr("variant_set", PackageNode, Variant).
 
 :- variant_is_propagated(PackageNode, Variant),
    attr("variant_selected", PackageNode, Variant, Value, _, _),
-   not propagate(PackageNode, variant_value(Variant, Value)).
-
-% a node is the source of propagation if nothing propagated to it
-propagation_source(Source, Variant, Value) :-
-  propagate(Source, variant_value(Variant, Value)),
-  not 1 { propagate(Parent, variant_value(Variant, Value)): depends_on(Parent, Source) }.
+   not propagate(PackageNode, variant_value(Variant, Value, _)).
 
 error(100, "{0} and {1} cannot both propagate variant '{2}' to the shared dependency: {3}",
   Package1, Package2, Variant, Dependency) :-
-  % Package1 and Package2 are source packages set to propagate Variant
-  propagation_source(node(X, Package1), Variant, Value1),
-  propagation_source(node(Y, Package2), Variant, Value2),
-  % Package1 and Package2 and their values are different
-  Package1 > Package2,
-  Value1 != Value2,
   % The variant is a singlevalued variant
   variant_single_value(node(X, Package1), Variant),
   % Dependency is trying to propagate Variant with different values and is not the source package
-  propagate(node(Z, Dependency), variant_value(Variant, Value1)),
-  propagate(node(Z, Dependency), variant_value(Variant, Value2)),
-  not propagation_source(node(Z, Dependency), Variant, Value1),
-  not propagation_source(node(Z, Dependency), Variant, Value2).
+  propagate(node(Z, Dependency), variant_value(Variant, Value1, node(X, Package1))),
+  propagate(node(Z, Dependency), variant_value(Variant, Value2, node(Y, Package2))),
+  % Package1 and Package2 and their values are different
+  Package1 > Package2,  Value1 != Value2,
+  not propagate(node(Z, Dependency), variant_value(Variant, _, node(Z, Dependency))).
 
 % Cannot propagate the same variant from two different packages if one is a dependency of the other
 error(100, "{0} and {1} cannot both propagate variant '{2}'", Package1, Package2, Variant) :-
-  % Package1 and Package2 are source packages set to propagate Variant
-  propagation_source(node(X, Package1), Variant, Value1),
-  propagation_source(node(Y, Package2), Variant, Value2),
-  % The variant is a singlevalued variant
+  % The variant is a single-valued variant
   variant_single_value(node(X, Package1), Variant),
   % Package1 and Package2 and their values are different
-  Package1 != Package2,
-  Value1 != Value2,
+  Package1 != Package2,  Value1 != Value2,
   % Package2 is set to propagate the value from Package1
-  propagate(node(Y, Package2), variant_value(Variant, Value1)),
+  propagate(node(Y, Package2), variant_value(Variant, Value2, node(X, Package2))),
+  propagate(node(Y, Package2), variant_value(Variant, Value1, node(X, Package1))),
   variant_is_propagated(node(Y, Package2), Variant).
 
 % Cannot propagate a variant if a different value was set for it in a dependency
 error(100, "Cannot propagate the variant '{0}' from the package: {1} because package: {2} is set to exclude it", Variant, Source, Package) :-
   % Package has a Variant and Source is propagating Variant
   attr("variant_set", node(X, Package), Variant, Value1),
-  propagation_source(node(Y, Source), Variant, Value2),
   % The packages and values are different
-  Source != Package,
-  Value1 != Value2,
-  % The variant is a singlevalued variant
+  Source != Package,  Value1 != Value2,
+  % The variant is a single-valued variant
   variant_single_value(node(X, Package1), Variant),
   % A different value is being propagated from somewhere else
-  propagate(node(X, Package), variant_value(Variant, Value2)).
+  propagate(node(X, Package), variant_value(Variant, Value2, node(Y, Source))).
 
 %----
 % Flags

```

alalazo replied:
https://github.com/spack/spack/pull/42931/commits/61b3ba3ceefbd914caf76af6d484816f313dcd0f

alalazo reviewed (commented):
Left a couple of comments.

RikkiButler20 committed:
Update variant check

Use the correct variant existence check post PR 44425

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented:
I'll do some benchmark of this branch tomorrow, and post here.

RikkiButler20 committed:
Dont check if variat was already set

alalazo commented:
This use case is not tested but, regardless of the behavior, I see a performance issue with it. I'm working on:

* **Spack:** 0.23.0.dev0 (8892c878ce173022c0efb18b35f1b21afd7a072b)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake

On this branch:
```console
$ time spack spec hdf5++shared ^zlib-ng~shared
Killed

real	6m31,216s
user	6m29,992s
sys	0m1,098s
```
while on `develop`:
```console
$ time spack spec hdf5++shared ^zlib-ng~shared
Input spec
--------------------------------
 -   hdf5++shared
 -       ^zlib-ng~shared

Concretized
--------------------------------
 -   hdf5@1.14.5%gcc@10.5.0~cxx~fortran~hl~ipo~java~map+mpi+shared~subfiling~szip~threadsafe+tools api=default build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-icelake
 -       ^cmake@3.30.5%gcc@10.5.0~doc+ncurses+ownlibs~qtgui build_system=generic build_type=Release patches=dbc3892 arch=linux-ubuntu20.04-icelake
 -           ^curl@8.10.1%gcc@10.5.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu20.04-icelake
 -               ^nghttp2@1.63.0%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -                   ^diffutils@3.10%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -               ^openssl@3.4.0%gcc@10.5.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-icelake
 -                   ^ca-certificates-mozilla@2023-05-30%gcc@10.5.0 build_system=generic arch=linux-ubuntu20.04-icelake
 -           ^ncurses@6.5%gcc@10.5.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-ubuntu20.04-icelake
 -       ^gcc-runtime@10.5.0%gcc@10.5.0 build_system=generic arch=linux-ubuntu20.04-icelake
[e]      ^glibc@2.31%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -       ^gmake@4.4.1%gcc@10.5.0~guile build_system=generic arch=linux-ubuntu20.04-icelake
 -       ^openmpi@5.0.5%gcc@10.5.0+atomics~cuda~debug~gpfs~internal-hwloc~internal-libevent~internal-pmix~java~lustre~memchecker~openshmem~romio+rsh~static+vt+wrapper-rpath build_system=autotools fabrics=none romio-filesystem=none schedulers=none arch=linux-ubuntu20.04-icelake
 -           ^autoconf@2.72%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -               ^m4@1.4.19%gcc@10.5.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-icelake
 -                   ^libsigsegv@2.14%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -           ^automake@1.16.5%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -           ^hwloc@2.11.1%gcc@10.5.0~cairo~cuda~gl~libudev+libxml2~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-ubuntu20.04-icelake
 -               ^libpciaccess@0.17%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -                   ^util-macros@1.20.1%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -               ^libxml2@2.10.3%gcc@10.5.0+pic~python+shared build_system=autotools arch=linux-ubuntu20.04-icelake
 -                   ^libiconv@1.17%gcc@10.5.0 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-icelake
 -                   ^xz@5.4.6%gcc@10.5.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu20.04-icelake
 -           ^libevent@2.1.12%gcc@10.5.0+openssl build_system=autotools arch=linux-ubuntu20.04-icelake
 -           ^libtool@2.4.7%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -               ^findutils@4.9.0%gcc@10.5.0 build_system=autotools patches=440b954 arch=linux-ubuntu20.04-icelake
 -           ^numactl@2.0.18%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -           ^openssh@9.8p1%gcc@10.5.0+gssapi build_system=autotools arch=linux-ubuntu20.04-icelake
 -               ^krb5@1.21.3%gcc@10.5.0+shared build_system=autotools arch=linux-ubuntu20.04-icelake
 -                   ^bison@3.8.2%gcc@10.5.0~color build_system=autotools arch=linux-ubuntu20.04-icelake
 -                   ^gettext@0.22.5%gcc@10.5.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu20.04-icelake
 -                       ^tar@1.34%gcc@10.5.0 build_system=autotools zip=pigz arch=linux-ubuntu20.04-icelake
 -                           ^pigz@2.8%gcc@10.5.0 build_system=makefile arch=linux-ubuntu20.04-icelake
 -                           ^zstd@1.5.6%gcc@10.5.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu20.04-icelake
 -               ^libedit@3.1-20240808%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -               ^libxcrypt@4.4.35%gcc@10.5.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu20.04-icelake
 -           ^perl@5.40.0%gcc@10.5.0+cpanm+opcode+open+shared+threads build_system=generic arch=linux-ubuntu20.04-icelake
 -               ^berkeley-db@18.1.40%gcc@10.5.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu20.04-icelake
 -               ^bzip2@1.0.8%gcc@10.5.0~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-icelake
 -               ^gdbm@1.23%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -                   ^readline@8.2%gcc@10.5.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-icelake
 -           ^pmix@5.0.3%gcc@10.5.0~munge~python~restful build_system=autotools arch=linux-ubuntu20.04-icelake
 -       ^pkgconf@2.2.0%gcc@10.5.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -       ^zlib-ng@2.2.1%gcc@10.5.0+compat+new_strategies+opt+pic~shared build_system=autotools arch=linux-ubuntu20.04-icelake


real	0m21,323s
user	0m20,226s
sys	0m0,986s
```
Looking into this because of https://github.com/spack/spack/pull/42931/commits/a82f58666418182127acf0d5e71338d6e8f27470, which now selects a propagated variant even if it was set on the node.

alalazo commented:
Ok, this would be solved by my diff above https://github.com/spack/spack/pull/42931#discussion_r1823521035, but the change in behavior wrt `develop` remains:
```console
$ time spack spec hdf5++shared ^zlib-ng~shared
==> Error: failed to concretize `hdf5++shared ^zlib-ng~shared` for the following reasons:
     1. Cannot propagate the variant 'shared' from the package: hdf5 because package: zlib-ng is set to exclude it

real	0m18,295s
user	0m17,373s
sys	0m0,843s
```

alalazo commented on lib/spack/spack/spec.py:
```diff
@@ -3362,7 +3367,16 @@ def satisfies(self, other: Union[str, "Spec"], deps: bool = True) -> bool:
+            # Check if any not satisified variant is propagating and in spec's dependency
+            if any(
+                other.variants[variant].propagate
+                and not self.variant_exists_in_dependency(variant)
```
@tgamblin This one couples the abstract spec semantic with packages, so the result varies depending on the repo:
```python
Spack version 0.23.0.dev0
Python 3.11.5, Linux x86_64
>>> import spack.spec
>>> s = spack.spec.Spec("hdf5++legacylaunchers")
>>> t = spack.spec.Spec("hdf5 ^openmpi")
>>> t.satisfies(s)
True
```
If we remove the variant from `openmpi` (or use another `mpi` provider), the same would be `False`. Are we fine with this semantic? Asking because up to know we tried hard not to couple things, so I'm wondering if this is deliberate, or accidental.

alalazo replied:
@RikkiButler20 @tgamblin This one I believe is a bug:
```python
Spack version 0.23.0.dev0
Python 3.11.5, Linux x86_64
>>> import spack.spec
>>> s = spack.spec.Spec("hdf5++legacylaunchers")
>>> t = spack.spec.Spec("hdf5 ^openmpi~legacylaunchers")
>>> t.satisfies(s)
True
```

alalazo replied:
Here's another big logical inconsistency:
```python
Spack version 0.23.0.dev0
Python 3.11.5, Linux x86_64
>>> import spack.spec
>>> s = spack.spec.Spec("hdf5++legacylaunchers")
>>> t = spack.spec.Spec("hdf5 ^openmpi~legacylaunchers")
>>> u = spack.spec.Spec("hdf5+legacylaunchers")
>>> s == u
True
>>> t.satisfies(s)
True
>>> t.satisfies(u)
False
```
`s` and `u` are equal, but `t` satisfies only one of them. This in my opinion is kind of dangerous, since we might `dedupe` stuff in containers based on `==` and get different results in `satisfies` calls depending which spec got in the container first.

EDIT: This `==` bug is already in `develop`

alalazo replied:
There's also https://github.com/spack/spack/issues/47347 on `develop`, but I'll try to deal with that separately

alalazo commented on var/spack/repos/builtin.mock/packages/openblas/package.py:
```diff
@@ -25,4 +25,6 @@ class Openblas(Package):
     # To ensure test works with newer gcc versions
     conflicts("%gcc@:10.1", when="@0.2.16:")
 
+    depends_on("perl")
```
Why we made the mock openblas depend on `perl`?

RikkiButler20 replied:
I used it for an example of propagating a variant that is not in the source package but is in a 1st and 2nd level dependencies. It was just to make sure the variant was propagated properly.

alalazo replied:
Ok, in that case I guess it would have been better to create another package with the relevant structure, and document briefly the graph/tree in the test docstring. 

That would document why we did what we did and, when some test fails in future PRs, would give some hint on the structure it was intended to test (so developers can check if anything changed in the meanwhile).

alalazo replied:
Anyhow, not a change requested in this PR.

alalazo committed:
Remove propagate_source/3 from the encoding

alalazo commented:
Trying to benchmark concretization on specs that _do not_ propagate variants, shows the same performance as develop, as expected.  The base for the tests is:

* **Spack:** 0.23.0.dev0 (8892c878ce173022c0efb18b35f1b21afd7a072b)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake

Results below for completeness.


<details>

<summary><b>radiuss performance benchmark (no propagation)</b></summary> 

[radiuss.develop.csv](https://github.com/user-attachments/files/17586681/radiuss.develop.csv)
[radiuss.pr.csv](https://github.com/user-attachments/files/17586682/radiuss.pr.csv)
[radiuss.txt](https://github.com/user-attachments/files/17586683/radiuss.txt)

![radiuss](https://github.com/user-attachments/assets/ee408edd-3c2c-4a2f-8001-febdb6f06c9a)


</details>

I also did a smaller performance benchmark on specs that do propagate variants, and measured the effect of https://github.com/spack/spack/pull/42931/commits/61b3ba3ceefbd914caf76af6d484816f313dcd0f

[propagate.develop.csv](https://github.com/user-attachments/files/17586708/propagate.develop.csv)
[propagate.pr.before_patch.csv](https://github.com/user-attachments/files/17586709/propagate.pr.before_patch.csv)
[propagate.pr.csv](https://github.com/user-attachments/files/17586710/propagate.pr.csv)
[benchmark.txt](https://github.com/user-attachments/files/17586711/benchmark.txt)

After the patch results are pretty much the same:

![propagate](https://github.com/user-attachments/assets/09138883-755c-4090-a1ac-de0ef9f3101f)

The effect of the patch on specs that wouldn't error is the following:

![propagate patch](https://github.com/user-attachments/assets/d2e458b7-892b-4a81-8324-eb40cd84a293)

alalazo committed:
Merge branch 'develop' into feature/propagate-variant-not-source

alalazo committed:
Remove Spec.satisfies dependency on spack.repo.PATH

This change makes `Spec.satisfies` independent of the details of
the repositories being used. This also changes the semantic slightly,
as now propagating variants will be satisfied if not contradicted.

For instance:

  spec.satisfies("++mpi")

if it has no `~mpi` variant set in any node. It follows that a spec
with no `mpi` variants satisfies `++mpi`.

alalazo committed:
Extend non-contradiction semantic to abstract specs

alalazo commented:
Here's the last benchmark, done using the _mean_ time of 10 repetitions + error bars:

* **Spack:** 0.23.0.dev0 (14bc900e9db7ddaaf32d79c32386a9c81f65c18b)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake

[propagate.develop.csv](https://github.com/user-attachments/files/17638899/propagate.develop.csv)
[propagate.pr.csv](https://github.com/user-attachments/files/17638901/propagate.pr.csv)
[propagation.txt](https://github.com/user-attachments/files/17638902/propagation.txt)

![radiuss](https://github.com/user-attachments/assets/ad60bb1a-31f4-4616-a3d7-ad9be5899a15)

Everything seems in line with `develop`.
